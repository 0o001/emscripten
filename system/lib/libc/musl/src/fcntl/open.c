#include <fcntl.h>
#include <stdarg.h>
#include "syscall.h"
#include "libc.h"

int open(const char *filename, int flags, ...)
{
	mode_t mode = 0;

	if ((flags & O_CREAT) || (flags & O_TMPFILE) == O_TMPFILE) {
		va_list ap;
		va_start(ap, flags);
		mode = va_arg(ap, mode_t);
		va_end(ap);
	}

#ifdef __EMSCRIPTEN__
	int fd = emscripten_path_open(filename, flags, mode);
	// CLOEXEC makes no sense for a single process
#else
	int fd = __sys_open_cp(filename, flags, mode);
	if (fd>=0 && (flags & O_CLOEXEC))
		__syscall(SYS_fcntl, fd, F_SETFD, FD_CLOEXEC);
#endif

	return __syscall_ret(fd);
}

LFS64(open);
