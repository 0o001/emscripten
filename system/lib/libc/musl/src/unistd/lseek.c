#include <unistd.h>
#include "syscall.h"
#include "libc.h"

off_t lseek(int fd, off_t offset, int whence)
{
#ifdef SYS__llseek
	off_t result;
#ifdef __EMSCRIPTEN__
	__wasi_errno_t err = __wasi_fd_seek(fd, offset, whence, &result);
  if (err != __WASI_ESUCCESS) {
		__wasi_syscall_ret(err);
		return -1;
	}
	return result;
#else
	return syscall(SYS__llseek, fd, offset>>32, offset, &result, whence) ? -1 : result;
#endif // __EMSCRIPTEN__
#else
	return syscall(SYS_lseek, fd, offset, whence);
#endif
}

LFS64(lseek);
