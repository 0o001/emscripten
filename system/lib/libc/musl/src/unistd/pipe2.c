#include <unistd.h>
#include <errno.h>
#include <fcntl.h>
#include "syscall.h"

int pipe2(int fd[2], int flag)
{
	if (!flag) return pipe(fd);
	int ret = __syscall(SYS_pipe2, fd, flag);
	if (ret != -ENOSYS) return __syscall_ret(ret);
	ret = pipe(fd);
	if (ret) return ret;
#ifndef __EMSCRIPTEN__ // CLOEXEC makes no sense for a single process
	if (flag & O_CLOEXEC) {
		__syscall(SYS_fcntl, fd[0], F_SETFD, FD_CLOEXEC);
		__syscall(SYS_fcntl, fd[1], F_SETFD, FD_CLOEXEC);
	}
#endif
	if (flag & O_NONBLOCK) {
#ifdef __EMSCRIPTEN__
     __wasi_fd_fdstat_set_flags(fd[0], __WASI_FDFLAG_NONBLOCK);
     __wasi_fd_fdstat_set_flags(fd[1], __WASI_FDFLAG_NONBLOCK);
#else
		__syscall(SYS_fcntl, fd[0], F_SETFL, O_NONBLOCK);
		__syscall(SYS_fcntl, fd[1], F_SETFL, O_NONBLOCK);
#endif
	}
	return 0;
}
