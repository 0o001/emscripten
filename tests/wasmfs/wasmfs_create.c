/*
 * Copyright 2021 The Emscripten Authors.  All rights reserved.
 * Emscripten is available under two separate licenses, the MIT license and the
 * University of Illinois/NCSA Open Source License.  Both these licenses can be
 * found in the LICENSE file.
 */

#include <assert.h>
#include <errno.h>
#include <fcntl.h>
#include <stdio.h>
#include <string.h>
#include <unistd.h>

// FIXME: Merge with other existing close and open tests.

int main() {
  // Test creating a new file and writing and reading from it.
  int fd = open("/test", O_RDWR | O_CREAT);
  const char* msg = "Test\n";
  write(fd, msg, strlen(msg));
  char buf[100];
  read(fd, buf, sizeof(buf));
  printf("%s", buf);
  close(fd);

  // Try to create an existing file with O_EXCL and O_CREAT
  errno = 0;
  int fd2 = open("/dev/stdin", O_RDWR | O_CREAT | O_EXCL);
  printf("Errno: %s\n", strerror(errno));
  assert(errno == EEXIST);

  // Try to open a file with O_DIRECTORY
  errno = 0;
  int fd3 = open("/dev/stdin", O_RDWR | O_DIRECTORY);
  printf("Errno: %s\n", strerror(errno));
  assert(errno == ENOTDIR);

  // Try to open a directory with O_DIRECTORY
  errno = 0;
  int fd4 = open("/dev", O_RDONLY | O_DIRECTORY);
  printf("Errno: %s\n", strerror(errno));
  assert(errno == 0);

  return 0;
}
