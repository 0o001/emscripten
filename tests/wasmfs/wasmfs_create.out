Test
Errno: File exists
Errno: Not a directory
Errno: No error information
